core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/UnwrapCastInComparison.java

```diff
@@ -172,6 +175,117 @@ public Expression rewriteComparisonExpression(ComparisonExpression node, Void co
             return unwrapCast(expression);
         }
 
+        @Override
+        public Expression rewriteBetweenPredicate(BetweenPredicate node, Void context, ExpressionTreeRewriter<Void> treeRewriter)
+        {
+            BetweenPredicate expression = (BetweenPredicate) treeRewriter.defaultRewrite((Expression) node, null);
+
+            if (!(expression.getValue() instanceof Cast cast)) {
+                return expression;
+            }
+
+            ComparisonExpression lowBound = new ComparisonExpression(GREATER_THAN_OR_EQUAL, node.getValue(), node.getMin());
+            ComparisonExpression highBound = new ComparisonExpression(LESS_THAN_OR_EQUAL, node.getValue(), node.getMax());
```

Comment:
Don't synthesize expression for these comparisons. Instead, the underlying methods should be adjusted to not have to take a ComparisonExpression. See https://github.com/trinodb/trino/pull/18138#discussion_r1253337523

Reply:
Reused logic from #18138 

Thank you @kabunchi